akka-stream-tests/src/test/scala/akka/stream/scaladsl/HubSpec.scala

```diff
@@ -0,0 +1,353 @@
+      val (sink, result) = MergeHub.source[Int](16).take(20000).toMat(Sink.seq)(Keep.both).run()
+      Source(1 to 10000).runWith(sink)
+      Source(10001 to 20000).runWith(sink)
+
+      result.futureValue.sorted should ===(1 to 20000)
+    }
+
+    "work with long streams when buffer size is 1" in assertAllStagesStopped {
+      val (sink, result) = MergeHub.source[Int](1).take(20000).toMat(Sink.seq)(Keep.both).run()
+      Source(1 to 10000).runWith(sink)
+      Source(10001 to 20000).runWith(sink)
+
+      result.futureValue.sorted should ===(1 to 20000)
+    }
+
+    "work with long streams when consumer is slower" in assertAllStagesStopped {
+      val (sink, result) =
+        MergeHub.source[Int](16)
+          .take(2000)
+          .throttle(10, 1.millisecond, 200, ThrottleMode.shaping)
+          .toMat(Sink.seq)(Keep.both)
+          .run()
+
+      Source(1 to 1000).runWith(sink)
+      Source(1001 to 2000).runWith(sink)
+
+      result.futureValue.sorted should ===(1 to 2000)
+    }
+
+    "work with long streams if one of the producers is slower" in assertAllStagesStopped {
+      val (sink, result) =
+        MergeHub.source[Int](16)
+          .take(2000)
+          .toMat(Sink.seq)(Keep.both)
+          .run()
+
+      Source(1 to 1000).throttle(10, 1.millisecond, 100, ThrottleMode.shaping).runWith(sink)
+      Source(1001 to 2000).runWith(sink)
+
+      result.futureValue.sorted should ===(1 to 2000)
+    }
+
+    "work with different producers separated over time" in assertAllStagesStopped {
+      val downstream = TestSubscriber.probe[immutable.Seq[Int]]()
+      val sink = MergeHub.source[Int](16).grouped(100).toMat(Sink.fromSubscriber(downstream))(Keep.left).run()
+
+      Source(1 to 100).runWith(sink)
+      downstream.requestNext() should ===(1 to 100)
+
+      Source(101 to 200).runWith(sink)
+      downstream.requestNext() should ===(101 to 200)
+
+      downstream.cancel()
+    }
+
+    "keep working even if one of the producers fail" in assertAllStagesStopped {
+      val (sink, result) = MergeHub.source[Int](16).take(10).toMat(Sink.seq)(Keep.both).run()
+      EventFilter.error("Upstream producer failed with exception").intercept {
+        Source.failed(TE("faling")).runWith(sink)
+        Source(1 to 10).runWith(sink)
+      }
+
+      result.futureValue.sorted should ===(1 to 10)
+
+    }
+
+  }
+
+  "BroadcastHub" must {
+
+    "work in the happy case" in assertAllStagesStopped {
+      val source = Source(1 to 10).runWith(BroadcastHub.sink(8))
+      source.runWith(Sink.seq).futureValue should ===(1 to 10)
+    }
+
+    "send the same elements to consumers attaching around the same time" in assertAllStagesStopped {
+      val (firstElem, source) = Source.maybe[Int].concat(Source(2 to 10)).toMat(BroadcastHub.sink(8))(Keep.both).run()
+
+      val f1 = source.runWith(Sink.seq)
+      val f2 = source.runWith(Sink.seq)
+
+      // TODO: Ensure subscription of Sinks
```

Comment:
Ah, ofc. its internally subscribed. Well then sleeping is the best we can do, so remove the todo markers?